openfl/utilities/fed_timer.py

```diff
@@ -0,0 +1,61 @@
+import logging
+import os
+import sys
+import time
+from threading import Thread
+
+logger = logging.getLogger(__name__)
+
+class CustomThread(Thread):
+    def __init__(self, group=None, target=None, name=None,
+                    args=(), kwargs={}):
+        Thread.__init__(self, group, target, name, args, kwargs)
+        self._result = None
+        self._duration = None
+
+    def run(self):
+        if self._target is not None:
+            start = time.time()
+            self._result = self._target(*self._args, **self._kwargs)
+            self._duration = round(time.time() - start, 3)
+
+    def result(self):
+        return self._result, self._duration
+
+def parameterize(dec):
+    def layer(*args, **kwargs):
+        def repl(f):
+            return dec(f, *args, **kwargs)
+        return repl
+    return layer        
+
+@parameterize
+def fedtiming(func, timeout):
+    def wrapper(*args, **kwargs):
+        start = time.time()
+        logger.info(f"Started: ({func.__name__})")
+
+        task = CustomThread(target=func, name=func.__name__, args=args, kwargs=kwargs)
+        task.start()
+        task.join(timeout)
```

Comment:
Does this mean that we will wait for the whole timeout and check the results only after that? So if a task is finished earlier we would not know?

Reply:
Not really.

For example, for any single instance of a decorated function

if the actual execution time is 2 seconds and the timeout value is 5 seconds then everything runs only for 2 seconds it doesn't wait until the timeout 5 seconds are over.
So in this case the total execution time would be `min(2, 5)`.


If the actual execution time exceeds the timeout value then the `task.join(timeout)` kicks in and we check if the thread is still running/alive then we raise an exception, log the
`TimeoutError` and terminate that thread.
So the max total execution time in this case is just capped by the timeout value i.e., '5 seconds'

Hope it helps.

Reply:
Got it, thank you!